lib/shared/src/event/log_event/mod.rs

```diff
@@ -0,0 +1,828 @@
+        match this_segment {
+            Segment::Coalesce(sub_segments) => {
+                // Creating a needle with a back out of the loop is very important.
+                let mut needle = None;
+                for sub_segment in sub_segments {
+                    let mut lookup = Lookup::try_from(sub_segment).ok()?;
+                    // Notice we cannot take multiple mutable borrows in a loop, so we must pay the
+                    // contains cost extra. It's super unfortunate, hopefully future work can solve this.
+                    lookup.extend(working_lookup.clone()); // We need to include the rest of the removal.
+                    if self.contains(lookup.clone()) {
+                        trace!(option = %lookup, "Found coalesce option.");
+                        needle = Some(lookup);
+                        break;
+                    } else {
+                        trace!(option = %lookup, "Did not find coalesce option.");
+                    }
+                }
+                match needle {
+                    Some(needle) => self.get(needle),
+                    None => None,
+                }
+            }
+            Segment::Field {
+                name,
+                requires_quoting: _,
+            } => {
+                if working_lookup.len() == 0 {
+                    // Terminus: We **must** insert here or abort.
+                    trace!(field = %name, "Getting from root.");
+                    self.fields.get(name)
+                } else {
+                    trace!(field = %name, "Descending into map.");
+                    match self.fields.get(name) {
+                        Some(v) => v.get(working_lookup).unwrap_or_else(|error| {
+                            debug!("{:?}", error);
+                            None
+                        }),
+                        None => None,
+                    }
+                }
+            }
+            // In this case, the user has passed us an invariant.
+            Segment::Index(_) => {
+                error!(
+                    "Lookups into LogEvents should never start with indexes.\
+                        Please report your config."
+                );
+                None
+            }
+        }
+    }
+
+    /// Get a mutable borrow of the value by lookup.
+    ///
+    /// ```rust
+    /// use shared::{log_event, event::*, lookup::*};
+    /// let plain_key = "foo";
+    /// let lookup_key = LookupBuf::from_str("bar.baz").unwrap();
+    /// let mut event = log_event! {
+    ///     plain_key => 1,
+    ///     lookup_key.clone() => 2,
+    /// }.into_log();
+    /// assert_eq!(event.get_mut(plain_key), Some(&mut Value::from(1)));
+    /// assert_eq!(event.get_mut(&lookup_key), Some(&mut Value::from(2)));
+    /// ```
+    pub fn get_mut<'a>(&mut self, lookup: impl Into<Lookup<'a>> + Debug) -> Option<&mut Value> {
```

Comment:
I absolutely agree, and I've been pondering how exactly to do it.

There's a few complications:

* It'd need to be generic over `&self` and `&mut self` if it were to work for get and get_mut/insert/remove
   + Having a `descent` and `descent_mut` option might be a cozy middle ground.
   + It would need to handle several possible cases, which might end up making it not much nicer than it is now.

Reply:
A few other reasons we might want `LogEvent` to hold a `Value`:

* In `get` and `get_mut` we could support `event.get(Lookup::from_str(".").unwrap())`, instead of hitting an `.expect`
* In `paths` we could return `(., self.fields)` instead of being unable to return this.

Reply:
That makes sense to me. The top-level is special in that we know which type it is, but that doesn't make it not a `Value`.

Reply:
Let me investigate doing that. :)

Reply:
Looks to be mostly working! I have a few doc tests to fix then I want to add tests for the cases I listed.

Reply:
Totally worked!